### PR TITLE
Update error message when lucene max document limit is breached

### PR DESCRIPTION
**Pull request by rayshrey**
### Description
Whenever lucene document limit(which is currently INT_MAX) is breached for any shard, the error message that is shown currently indicates that the limit is on an index level whereas actually the limit is on a shard level.

Updating the error message in this PR to indicate the same. Also including the Shard Id as well in the message.

Previous message
```
Number of documents in the index can't exceed [2147483647]
```

Updated message
```
Number of documents in shard <index_name>[<shardId>] exceeds the limit of [2147483647] documents per shard
```

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List

- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [ ] ~New functionality has been documented.~
 - [ ] ~New functionality has javadoc added~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -1412,7 +1412,9 @@ private Exception tryAcquireInFlightDocs(Operation operation, int addingDocs) {
         final long totalDocs = indexWriter.getPendingNumDocs() + inFlightDocCount.addAndGet(addingDocs);
         if (totalDocs > maxDocs) {
             releaseInFlightDocs(addingDocs);
-            return new IllegalArgumentException("Number of documents in the index can't exceed [" + maxDocs + "]");
+            return new IllegalArgumentException(
+                "Number of documents in single shard of index can't exceed [" + maxDocs + "] - " + shardId.toString()
```
Comment (mgodwan) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1416]: We don't need to invoke `toString`

Comment (mgodwan) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1416]:
```suggestion
                String.format("Number of documents in the single shard [%s] can't exceed [%s]", shardId, maxDocs)
```

**Review by mgodwan — CHANGES_REQUESTED** [review]

**Review by kkmr — APPROVED** [review]
Curious to know why we need this change?

**Comment by dblock** [comment]
Let's not skip CHANGELOG here, I removed the label.

What does one do when they receive this message? Is there a different max number of documents per index that can be inferred from this and included in the error? Just trying to think about how actionable this message is for the user and whether it should contain more information.

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -1412,7 +1412,7 @@ private Exception tryAcquireInFlightDocs(Operation operation, int addingDocs) {
         final long totalDocs = indexWriter.getPendingNumDocs() + inFlightDocCount.addAndGet(addingDocs);
         if (totalDocs > maxDocs) {
             releaseInFlightDocs(addingDocs);
-            return new IllegalArgumentException("Number of documents in the index can't exceed [" + maxDocs + "]");
+            return new IllegalArgumentException("Number of documents in shard " + shardId + " can't exceed [" + maxDocs + "]");
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1415]: I think we are not delivering the message here we actually, it gives impression that there is a problem with one specific shard but the limit applies to all shards of the index. I would frame it like that:
```
return new IllegalArgumentException("Number of the documents in shard " + shardId + " exceeds the limit of [" + maxDocs + "] per index shard");
```

Comment (rayshrey) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java]: agree @reta 
Update the error message to reflect the same 
```
"Number of documents in shard " + shardId + " exceeds the limit of [" + maxDocs + "] documents per shard"
```

**Comment by dblock** [comment]
@reta good?

**Review by kkmr — APPROVED** [review]

**Comment by reta** [comment]
@mgodwan LGTM?

**Review by mgodwan — APPROVED** [review]

**Comment by dblock** [comment]
@rayshrey you'll unfortunately need to manually backport to 2.x if you want this in the next 2.x release

**Comment by rayshrey** [comment]
@dblock raised the manual backport here - #11382

**Pull request merged**